piet-direct2d/src/lib.rs

```diff
@@ -555,20 +559,372 @@ impl<'a> TextLayoutBuilder for D2DTextLayoutBuilder<'a> {
+            metrics: HitTestMetrics {
+                text_position,
+                is_text: htp.metrics.is_text(),
+            },
+            is_inside: htp.is_inside,
+        }
+    }
+
+    /// Hit Test for Text Position.
+    ///
+    /// Given a text position (as a utf-8 code unit), returns the `x` offset of the associated grapheme cluster (generally).
+    ///
+    /// Can panic if text position is not at a code point boundary, or if it's out of bounds.
+    fn hit_test_text_position(&self, text_position: usize, trailing: bool) -> Option<HitTestTextPosition> {
+        // trailing not supported currently; waiting for BIDI
+        if trailing {
+            return None;
+        };
+
+        // Note: Directwrite will just return the line width if text position is
+        // out of bounds. This is what want for piet; return line width for the last text position
+        // (equal to line.len()). This is basically returning line width for the last cursor
+        // position.
+
+        // Now convert the utf8 index to utf16.
+        // This can panic;
+        let idx_16 = count_utf16(&self.text[0..text_position]);
+
+        // panic or Result are also fine options for dealing with overflow. Using Option here
+        // because it's already present and convenient.
+        // TODO this should probably go before convertin to utf16, since that's relatively slow
+        let idx_16 = idx_16.try_into().ok()?;
+
+        // TODO quick fix until directwrite fixes bool bug
+        let trailing = !trailing;
+
+        self.layout.hit_test_text_position(idx_16, trailing)
+            .map(|http| {
+                HitTestTextPosition {
+                    point: Point {
+                        x: http.point_x as f64,
+                        y: http.point_y as f64,
+                    },
+                    metrics: HitTestMetrics {
+                        text_position: text_position, // no need to use directwrite return value
+                        is_text: http.metrics.is_text(),
+                    },
+                }
+            })
+    }
+}
+
+/// Counts the number of utf-16 code units in the given string.
+/// from xi-editor
+pub(crate) fn count_utf16(s: &str) -> usize {
+    let mut utf16_count = 0;
+    for &b in s.as_bytes() {
+        if (b as i8) >= -0x40 {
+            utf16_count += 1;
+        }
+        if b >= 0xf0 {
+            utf16_count += 1;
+        }
+    }
+    utf16_count
+}
+
+/// returns utf8 text position (code unit offset)
+/// at the given utf-16 text position
+pub(crate) fn count_until_utf16(s: &str, utf16_text_position: usize) -> Option<usize> {
+    let mut utf8_count = 0;
+    let mut utf16_count = 0;
+    for &b in s.as_bytes() {
+        if (b as i8) >= -0x40 {
+            utf16_count += 1;
+        }
+        if b >= 0xf0 {
+            utf16_count += 1;
+        }
+
+        // When count goes beyond text position, it means the start boundary of the utf16 code unit is passed.
```

Comment:
I think this comment can be removed; it no longer accurately reflects the code, and didn't help me greatly with understanding.